### ISSUE TITLE
<Prompt /> component from react-router-dom misbehaves on cancelling navigation 

### ISSUE THREAD

**Issue by ukhancc** [remix-run/react-router#5405]
I've recently migrated from version 3 to 4.1.2 in the hope that the following bug caused by the `setRouteLeaveHook` won't be present in the `<Prompt />` component from `react-router-dom`.

So, as per the use case of `Prompt` component, when the user standing in the form refreshes the page, edit some fields and clicks back button, the confirmation alert pops up and when the user cancels the navigation, it stays on the page but the url gets changed and does not remain the same as of the current form page.

**Comment by timdorr** [comment]
Can you post a more concrete example? I'm not seeing this behavior myself. Are you using a HashRouter, by chance?

**Comment by arvinsingla** [comment]
I'm experiencing the exact same thing using `BrowserRouter`. If the user clicks a link or does an operation that pushes a route to history everything works as expected. If however the user clicks the browser back button I get a prompt as expected but the URL in the address bar has changed to the previous path.

If I then click cancel on the prompt indicating I do not wish to leave, one of two things will happen.

1. The current route is restored in the address bar
2. The previous route stays in the address bar

In both cases my current component is still shown and my components lifecycle hooks including`componentWillReceiveProps` are called which messes up the state of my component.

**Comment by zhouzefei** [comment]
i have the same question. have you resolved?

**Comment by ukhancc** [comment]
@timdorr just FYI, I'm using the `BrowserRouter` and my issue is exactly the same as @arvinsingla described "In both cases my current component is still shown and my components lifecycle hooks includingcomponentWillReceiveProps are called which messes up the state of my component."

**Comment by pshrmn** [comment]
If any of you are diving into this, PRs are always appreciated for this sort of thing.

At the very least, someone should provide a minimal reproduction repo.  It is a lot easier for other people to look into this issue if they have code they can take a look at. Just describing a problem means that anyone who wants to help has to first build their own project before they can attempt to debug, which isn't exactly motivating.

If there actually is an issue with the code here (I'd have to see it happen. I can't think of a reason why what is being described is happening unless there are browser related issues), then it is most likely going to be an issue with `history`, not `react-router`.

**Comment by timdorr** [comment]
Also, to make things easier, you can use our codesandbox project as a starting point: https://codesandbox.io/s/n55VljYk7

**Comment by pshrmn** [comment]
CodeSandbox unfortunately won't work here. Their fake browser implementation doesn't work correctly with the `<Prompt>`. I mentioned it to one of the devs, so they are at least somewhat aware of that issue, but I would consider that bug pretty low priority. That wouldn't be the same issue that is happening here, but it might give the illusion of it.

**Comment by timdorr** [comment]
Ah, I thought they passed that through to the browser. My bad.

**Comment by pshrmn** [comment]
It has some browser integration, but I wish that it didn't. I have some embedded sandboxes in the curi docs website and its annoying that the back button is hijacked. I much prefer Ryan's fake browser that uses an in memory history.

Back to this issue, I was trying to think of why `componentWillReceiveProps` is being triggered, and there might be an issue there if the navigation was done using the browser's forward/back button. Those are caught by event listeners, and history attempts to reverse the event (counter `go(-1)` with `go(1)`). `history` also will notify any listeners when this happens, but maybe it should be doing nothing when that happens?

**Comment by rahildar** [comment]
i was checking how to use Prompt but it suddenly stopped working 
https://reacttraining.com/react-router/web/example/preventing-transitions
couldn't figure out what was wrong , here is a video link https://drive.google.com/file/d/0B3SfWVzqRpFrdUpzN19pQ001WHc/view

**Comment by alexandersoto** [comment]
I'm running into this this issue as well. I've created a a test case: https://codesandbox.io/s/4lvjoow5mw. I get the same behavior on a real browser.

**Comment by pshrmn** [comment]
@alexandersoto I loaded your example as a full page (https://4lvjoow5mw.codesandbox.io/) and it worked as expected (although you seem to need to load that page directly, there is some weird behavior if you open it from the editor page).

**Comment by alexandersoto** [comment]
@pshrmn thanks for checking it out so quickly. The test case appears to work as a full page, but I found another related issue. Click on Broken -> go back -> go back a second time. The dialog box disappears and the url changes to `/`, but the router is rendering `/broken`. 

Going back a second time cancels the dialog, but doesn't update the url like clicking cancel does. I'm testing this out on Chrome.

**Comment by pshrmn** [comment]
Hmm, that second case is interesting. The behavior is different between Chrome and Firefox (Firefox will go to the page two pages back).
```js
// given the history
const history = [
  'google.com',
  'sandbox.com',
  'sandbox.com/block'
];
let index = 2;
```
In Firefox, clicking the back button once will open up the prompt. Clicking it a second time (while the prompt is open) will go to 'google.com'. In Chrome, as you described above, the first click is the same, but the second click will go to 'sandbox.com'. The Firefox behavior is what I would expect to happen.

I also forked your sandbox (https://codesandbox.io/s/l248ml7kj7) to add a third route so I could see what the behavior is when I can go back twice within the same site. When leaving the site entirely, that still works, but in both Chrome and Firefox it has weird behavior because it stacks prompts.

This double prompt issue is actually with the [history](https://github.com/ReactTraining/history) package, so it would probably be best to work on it over there.

As for the original issue from here, we still need a reproducible test case, so if anyone wants to put one together, it would be appreciated.

**Comment by Stashevich** [comment]
I have a same issue with <Prompt>.. Sometimes after clicking browser back button the hash url is returned back, if I say "do not redirect me", sometimes no..

**Comment by joaquindk** [comment]
Same issue here as @arvinsingla described - I'm experiencing the exact same two scenarios, and they both seem to occur quite randomly.

**Comment by zxlin** [comment]
I made a gif of the issue I'm experiencing, which is rather strange behavior: https://gfycat.com/HotNeatInganue

Repro steps:
1. load the codesandbox example from above (https://4lvjoow5mw.codesandbox.io)
2. click on the link "Broken"
3. click back (browser back)
4. notice how the path in the address bar has already changed to `/`
4. click on "cancel" in the prompt
5. notice how the path in the address bar changes back to `/broken`

The path change between steps 3 and 4 makes React think you've already navigated away before you even get a chance to select "Cancel" or "Ok" on the dialog box, so React unmounts the "Broken" component and then re-renders the "Broken" component after you click on "Cancel" which is when the path switches back. So if you're using `<Prompt />` on a page with a form, then you just lost all the data saved in the form component's state due to the unmount.

**Comment by neilpalima** [comment]
Is there any hack for this to be solve in the mean time?

**Comment by Edistra** [comment]
I experienced the same issue.

When I come from an in app route, the behaviour of Prompt is working as expected, but when I access the page directly with the url of the view containing the Prompt component and I trigger it, the browser url remains modified even if I click on cancel.

I made a repo to reproduce the issue : https://github.com/Edistra/react-router-dom-prompt-issue

I also made a gif of the issue :

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/13304634/37617051-48d9735e-2b88-11e8-8602-e84375f70595.gif)

**Comment by ekilah** [comment]
I seem to be observing similar behavior without the `<Prompt>` component, but instead while trying to implement something like https://github.com/ReactTraining/react-router/issues/4635#issuecomment-300465164 (a custom styled popup in my app to handle the back button). The URL bar string changes before my custom popup renders, but the page acts as expected (my popup shows and the app doesn't re-route). The difference is it always seems to change & stay changed after using the browser back button, rather than the reported cases here where the URL flips back after the dialog is closed.

**Comment by Piemontez** [comment]
I have the same problem
Look this animation.
![out-2](https://user-images.githubusercontent.com/1814990/39649798-bbdcb386-4fbc-11e8-9a11-9644b8489209.gif)

Below this part of my code

```
import React, {Component} from 'react';
import { Prompt } from 'react-router-dom';
...
class AgricultorEdit  extends Component {
  ...
  render() {
    return (
      <div>
        <Prompt
            when={!!props.change}
            message={location => 'Você alterou o cadastro e não salvou, tem certeza que deseja sair desta tela?' } />
        ....
      </div>
    )
  }
```

**Comment by Fabianopb** [comment]
Any updates on this issue @timdorr ? I'm experiencing the same problem, using React Router v4, React Router Redux 5-alpha, and using a pretty similar solution to [this one](https://github.com/ReactTraining/react-router/issues/4635#issuecomment-297828995) to show a custom modal.

**Apparently the issue happens when entering the route via POP action** (i.e. by directly accessing/reloading), but not if you come via PUSH (navigating from another route in the app).

Could this be related to the `history` api instead (https://github.com/ReactTraining/history#blocking-transitions)?

**Comment by thomasthiebaud** [comment]
Any update on this issue ? I'm facing the same problem

**Comment by CuiZhengyang** [comment]
Any update on this issue ? I'm facing the same problem too

**Comment by tobloef** [comment]
This bug is over a year old and one of the most upvoted issues here. It would be very nice to get an update on the status of this issue or some hack to fix this. Thanks!

**Comment by davidxmoody** [comment]
In case this is helpful to anyone, I ended up writing my own version of the Prompt component which doesn't have this bug and also prompts for leaving the page by closing the tab. Written in TypeScript and working with `history/createBrowserHistory` from `history@4.7.2`, `react-router-dom@4.3.1` and `react-router-redux@5.0.0-alpha.8`.

```typescript
import * as React from "react"
import {withRouter} from "react-router"

// Explanation: This component is meant to prevent accidental page transitions
// which could cause unsaved work to be lost. When this component is mounted
// and the `when` prop is true, it does two things:
//
// 1. Ask for confirmation before allowing page transitions from back/forward
//    browser navigation and react-router Link navigation. This uses the
//    react-router provided history.
// 2. Ask for confirmation before allowing "leave page" browser navigation
//    (close/refresh). This uses the window's beforeunload event.
//
// The first one is what the react-router Prompt component is supposed to do.
// However, it is buggy. See here:
// https://github.com/ReactTraining/react-router/issues/5405

const DEFAULT_MESSAGE =
  "You have unsaved changes, are you sure you want to leave?"

interface Props {
  when: boolean
  message?: string
  history?: any
}

class NavigationLock extends React.PureComponent<Props> {
  public unblock?: () => void

  public componentDidMount() {
    if (this.props.when) {
      this.startBlocking()
    }
  }

  public componentWillUnmount() {
    this.stopBlocking()
  }

  public componentWillReceiveProps(nextProps: Props) {
    if (nextProps.when && !this.props.when) {
      this.startBlocking()
    } else if (!nextProps.when && this.props.when) {
      this.stopBlocking()
    }
  }

  public render() {
    return null
  }

  private onBeforeUnload = (event: any) => {
    // Prompts the user before closing the page, see:
    // https://developer.mozilla.org/en-US/docs/Web/Events/beforeunload
    event.preventDefault()
    event.returnValue = ""
  }

  private startBlocking = () => {
    if (!this.unblock) {
      const message = this.props.message || DEFAULT_MESSAGE
      this.unblock = this.props.history.block(message)
    }
    window.addEventListener("beforeunload", this.onBeforeUnload)
  }

  private stopBlocking = () => {
    if (this.unblock) {
      this.unblock()
      this.unblock = undefined
    }
    window.removeEventListener("beforeunload", this.onBeforeUnload)
  }
}

export default withRouter(NavigationLock)
```

**Comment by sysleaf** [comment]
Any update on this from official mentainer.

**Comment by jombie** [comment]
@davidxmoody Above code does not work when hitting back button. The route url changes and reverts back on cancel and the whole component is reloaded loosing the state, but the direct browser refresh works.

**Comment by nioposiquit** [comment]
Same problem on react-router-dom "^4.3.1" I am using.. I use Prompt to alert user when leaving the current page wherein data/inputs provided are not yet submitted. Is there any fix this time? Thanks.

**Comment by ifier** [comment]
The same with `"react-router-dom": "^5.0.0"` and without hashes.
![Untitled](https://user-images.githubusercontent.com/8005910/57082168-e3273000-6cfe-11e9-88b3-e9385c228f49.gif)

**Comment by marton-levay** [comment]
Ohai

I don't know whether this helps anyone out, but I had the same issue: On back button I was prompted but the url indeed changed and sometimes it messed up the page sometimes it did not. (react-router/ -dom version 5.0.0)

After some debugging I realized that I have been using components from 'react-router' and 'react-router-dom' completely mixed, sometimes I import them from one sometimes the other. Now this should not matter much, but apparently in this case it does.

After I have ditched 'react-router' in favor of importing everything from 'react-router-dom' it works as 'expected' (Such as on browser back button it : 1. change the url, but does not unload the component, 2. shows the prompt window, 3. if I choose cancel the url is written back and the component is not unmounted or messed up).

I'm also using `import {BrowserRouter as Router, Route} from 'react-router-dom';` with exported history `import {createBrowserHistory} from "history";
export default createBrowserHistory();`

**Comment by MskShahrukh** [comment]
I am having this very same issue, i have Formik implemented with react and the browser back button when clicked does throw the alert.. but upon clicking cancel button the state of the form is lost...
hence, doesnt make any sense if you prompt user for unsaved changes and just clean the form when they click cancel . :/
Im surprised to see that this issue is **open since 2017** , have tried all the upper mentioned tricks.. none of them work for me..

**Comment by goodwin64** [comment]
Are there any updates on this issue?

**Comment by shantp** [comment]
Is it possible to cancel the location change if the user clicks cancel? I too am experiencing the bug where the user cancels and the prompt does not show up a 2nd time and they lose form data even if they clicked cancel.

**Comment by sibelius** [comment]
try this `usePrompt` hook https://twitter.com/sseraphini/status/1123321881942265856

https://t.co/aJYtv8vYbi?amp=1

**Comment by shantp** [comment]
@sibelius Thank you I will try this.  What is `useRouter` referring to in that gist?

**Comment by sibelius** [comment]
```jsx
import { useContext } from 'react';
import { __RouterContext, RouteComponentProps } from 'react-router-dom';

export const useRouter = <TParams = {}>() => {
  return useContext(__RouterContext) as RouteComponentProps<TParams>;
};
```

**Comment by shantp** [comment]
@sibelius What is the correct way to use the usePrompt hook?

**Comment by shantp** [comment]
@sibelius What is the double underscore in front of RouterContext? Is that a TypeScript thing? When I try to use it, it tells me there's no RouterContext in react-router-dom.

**Comment by MeiKatz** [comment]
@shantp This is a naming convention we use in React Router to make clear, that it is **not** public api and it could change anytime without announcing it.

**Comment by shantp** [comment]
Thanks @MeiKatz, also it is only in React Router 5 and I was using 4 so I had to upgrade to access it.

**Comment by redbmk** [comment]
I've done a little more debugging and hopefully can shed some light on the unexpected behavior, though I still don't have a solution. Testing the default `Prompt` from `react-router-dom`, I'm seeing the same behavior in both Firefox 69.0.1 and Chrome 77.0.3865.90.

Let's say I navigate from page A -> B -> C, where only page C has a `Prompt` on it.

When I click the browser's back button, the URL changes to B and the prompt comes up. If I cancel, the URL changes back to C and data is retained. If I instead click OK, the URL stays at B and data changes accordingly.

**Here's where the issues begin:**
Now, after navigating to page C, if I **_refresh the page_**, the browser still has pages A and B in its history. If I click the browser's back button, the URL changes to B and I get the prompt, as expected. If I click OK, the URL stays at B and the data changes, also as expected. However, if I click cancel, the URL **_still_** stays at B, and data **_does not change_** to reflect the new URL.

FWIW the same also applies to the browser's Forward button.

Basically I think the key to finding unexpected behavior is when react-router didn't control the initial route navigation (e.g. when the URL was manually entered, the page was refreshed, or possibly from an `a` tag rather than a `Link`).

**Comment by emadalam** [comment]
For those still finding a workaround with this issue, the following seems to do the trick, a bit hacky though, but works 🤷‍♂ 

1. Block the history transition `history.block`
2. Listen to history changes using the react router dom's `history` object and use `window.history.replaceState` to replace the current history state when the changed location is not equal to the current location.

Hook:
```js
import {useCallback, useEffect} from 'react'
import {useHistory} from 'react-router-dom'

const noop = () => {}
const createURLFromLocation = location =>
  `${location.pathname}${location.search}`

const useNavigationPrompt = (when, message) => {
  const history = useHistory()

  useEffect(() => {
    let unBlock = noop
    let unListen = noop
    if (when) {
      unBlock = history.block(message)
      unListen = history.listen((location, action) => {
        const currentUrl = createURLFromLocation(history.location)
        const newUrl = createURLFromLocation(location)
        if (currentUrl !== newUrl) {
          window.history.replaceState(null, null, currentUrl)
        }
      })
    }
    return () => {
      unBlock()
      unListen()
    }
  })
}

export default useNavigationPrompt
```

Usage:
```js
const Form = (props) => {
  const [isDirty, setIsDirty] = useState(false)

  useNavigationPrompt(isDirty, 'Are you sure?')

  // ...

  <form>My Form or any other component</form>
}

```

**Comment by zxlin** [comment]
This is not yet fixed, can someone tag it fresh please. Thanks!

**Comment by edi** [comment]
It's weird that this has been open open for 2 years now, but with no actual solution, on an actively maintained repo. Is there something preventing us from fixing it ?

**Comment by timdorr** [comment]
Someone just needs to make a PR.

**Comment by onlyann** [comment]
This issue is still present with react-router 5.1.2.

Reading through the thread, I can't tell if the issue is with react router or the history package.

What is the best work-around so far, if any?

**Comment by MskShahrukh** [comment]
This is a browser related issue, not a issue in react-router or react itself. Not even related to JS.
Since what we are trying to achieve is to handle a browser event (back button click), during this the state of page is lost as the history array is popped then pushed again.

You can actually see this by implementing a window listener for 'onBeforeUnload'.. u will see it changing in the console. But notice only for back button not window close button. Which is strange.

I also went a little deeper to see how WordPress handles this situation, on WordPress u can also see a kind of jerk or delay in re-fetch of data when u click back button, it seems like WordPress developers have gone out of the way to maintain a seperate trail of changes of the form area in the browser's local storage.

So, its better to inform chrome, Firefox, etc about this so that they can patch it.

**Comment by onlyann** [comment]
Thank you @MskShahrukh .

If that can be of help to anyone, I have the following hack for [mobx-react-router](https://github.com/alisd23/mobx-react-router) to work around this behaviour:

```typescript
(MobxRouterStore.prototype as any)._updateLocation = function (location: Location<any>) {
  // Here be dragons...
  // Stop the same location from being pushed twice 
  // when clicking the browser's back button and using a Prompt
  if (location.key && location.key === this.location?.key) {
    return;
  }
  this.location = location;
}
```

**Comment by joaquinwojcik** [comment]
This is the only way I've found to prompt user every time he tries to exit and also keep the URL (it change to previous URL but if user press "cancel" it'll change again to current url)

```js
/* [...] */
const [callAccepted, setCallAccepted] = useState(false);

/* I omitted the function where I set callAccepted to true to keep it short */ 

const check = (location, action) => {
    return callAccepted
      ? "Are you sure you want to exit?"
      : true;
};

return (
<>
    <Prompt when={callAccepted} message={check} />
    <div>
     {/* rest of component here... */}
    </div>
</>
)
```

If I remove `when` prop from `<Prompt>` or even if I remove params  from `check` function, it'll show the prompt only one time without keeping the URL.

**Comment by SufiyaanRajput** [comment]
> This is the only way I've found to prompt user every time he tries to exit and also keep the URL (it change to previous URL but if user press "cancel" it'll change again to current url)
> 
> ```js
> /* [...] */
> const [callAccepted, setCallAccepted] = useState(false);
> 
> /* I omitted the function where I set callAccepted to true to keep it short */ 
> 
> const check = (location, action) => {
>     return callAccepted
>       ? "Are you sure you want to exit?"
>       : true;
> };
> 
> return (
> <>
>     <Prompt when={callAccepted} message={check} />
>     <div>
>      {/* rest of component here... */}
>     </div>
> </>
> )
> ```
> 
> If I remove `when` prop from `<Prompt>` or even if I remove params from `check` function, it'll show the prompt only one time without keeping the URL.

sss

**Comment by SufiyaanRajput** [comment]
> This is the only way I've found to prompt user every time he tries to exit and also keep the URL (it change to previous URL but if user press "cancel" it'll change again to current url)
> 
> ```js
> /* [...] */
> const [callAccepted, setCallAccepted] = useState(false);
> 
> /* I omitted the function where I set callAccepted to true to keep it short */ 
> 
> const check = (location, action) => {
>     return callAccepted
>       ? "Are you sure you want to exit?"
>       : true;
> };
> 
> return (
> <>
>     <Prompt when={callAccepted} message={check} />
>     <div>
>      {/* rest of component here... */}
>     </div>
> </>
> )
> ```
> 
> If I remove `when` prop from `<Prompt>` or even if I remove params from `check` function, it'll show the prompt only one time without keeping the URL.

@joaquinwojcik I have tried it and it does prevent me from a redirect whenever the condition is true. However the url is still updated to the previous one.

After a bit of tweaking, I have replaced the window history for now with the current path and now it doesn't change to the previous url.

`<Prompt when={true} message={(location, action) => {
        window.history.replaceState(null, null, "/your/current/path")

        return true
              ? "Are you sure you want to exit?"
              : true;
        }} />`

This fixes it for me now. It also has a support for majority of the browsers. People reading this may check the browser support as per the project's user base.

**Comment by joaquinwojcik** [comment]
> This fixes it for me now. It also has a support for majority of the browsers. People reading this may check the browser support as per the project's user base.

@SufiyaanRajput which browser are you using? I've tested it on Chrome ^84

**Comment by SufiyaanRajput** [comment]
@joaquinwojcik Version 84.0.4147.125 (Official Build) (64-bit) chrome

**Comment by wosephjeber** [comment]
I've been wrestling with the same issue, and finally found a solution that works for my use case. As others have pointed out, when a `POP` action triggers the `getUserConfirmation` method and the navigation is canceled, React Router correctly prevents the route transition within the React app, but the browser has already popped that item from `window.history`, so the URL in the browser will have changed.

My solution was to update my `getUserConfirmation` handler to detect a `POP` action and call `window.history.forward()` to return to the original URL within the browser history. But in order for `getUserConfirmation` to know what action was performed, I needed to pass along that information in the message from the `<Prompt />` component. The message is just a string, so I'm using a serialized JSON string to pass along the action along with my message and some other structured content that I need for my customized confirmation prompt.

As a bit of context, I'm using the `HashRouter` because my use case is within a browser extension. I can't verify whether the same issue happens or whether my solution is adequate for the `BrowserRouter`. I also don't know if this addresses all of the scenarios discussed in this thread, but I wanted to provide a reduced example of my solution in hopes that it can help someone out:

```jsx
// In the component that renders the <Prompt />:
<Prompt
  message={(_nextLocation, action) => {
    return JSON.stringify({
      action,
      message: 'Some message'
    })
  }}
/>

// In the component that renders the router:
<HashRouter
  getUserConfirmation={(payload, callback) => {
    const { action, message } = JSON.parse(payload);

    const confirmed = window.confirm(message)
    
    callback(confirmed)

    if (!confirmed && action === 'POP') window.history.forward()
  }}
>
  ...
</HashRouter />
```

**Comment by Andrew1431** [comment]
@wosephjeber Your solution worked wonders for me! Well I have built my own component that uses this prompt so my code was a bit different:

```typescript
  const onMessage = useCallback((location: any, action: any) => {
    if (action === 'POP') {
      window.history.forward();
    }
    const card = document.getElementById('dirty_save_revert');
    card?.classList.add('shake');
    return false;
  }, []);
```

This one just prevents the back automatically, (or other routes) and shakes a "Save / Cancel" popup. Works wonders!

It's the same component that Discord uses: When you change some settings on a server a pop-up shows, and won't let you close the tab/nav out until you revert your changes or save. It's quite a nice UX imo.

**Comment by payapula** [comment]
@Andrew1431 Thanks for your suggestion. Hope this solution help some developers who are struggling with the same issue.

The Problem we faced is explained neatly by @redbmk's [Comment](https://github.com/ReactTraining/react-router/issues/5405#issuecomment-533277464).

The issue is that the prompt is allowing the URL to change even after clicking on cancel navigation. This **happens only after you reload on a page that has prompt**.

We ended up using the following approach according to our prompt configuration.

```js
const currentPath = useRef(null);
 useEffect(() => {
        if (!!currentPath) {
            currentPath.current = window.location.pathname;
        }
 }, []);

 const handleBlockedNavigation = (nextLocation, action) => {
        if (formIsDirty) //A boolean state to check if the page has any unsaved changes
         {
            //A workaround for Prompt replacing the URL after page reload, even if navigation is cancelled
            if (currentPath?.current !== nextLocation.pathname && action === 'POP') {
                window.history.forward();
            }
            return false;
        }
        return true;
    };

 <Prompt when={blockNavigation} message={handleBlockedNavigation} />
```

**blockNavigation**  --- is the boolean state to trigger the Prompt. 

**handleBlockedNavigation** --- will check for the unsaved changes on the page and return either true or false. If it returns true then the navigation is allowed, if false then the navigation is blocked.

**currentPath** --- Ref holds the current URL and would be checked before forwarding the URL by _window.history.forward();_

You may find **blockNavigation**  and **formIsDirty** are doing same work (or think they may be redundant), but it is according to our page requirements, we are checking another boolean on the page to make sure the user is safe to navigate or not (Omitted other conditions around it for brevity).

**Comment by fachryadhitya** [comment]
> @wosephjeber Your solution worked wonders for me! Well I have built my own component that uses this prompt so my code was a bit different:
> 
> ```ts
>   const onMessage = useCallback((location: any, action: any) => {
>     if (action === 'POP') {
>       window.history.forward();
>     }
>     const card = document.getElementById('dirty_save_revert');
>     card?.classList.add('shake');
>     return false;
>   }, []);
> ```
> 
> This one just prevents the back automatically, (or other routes) and shakes a "Save / Cancel" popup. Works wonders!
> 
> It's the same component that Discord uses: When you change some settings on a server a pop-up shows, and won't let you close the tab/nav out until you revert your changes or save. It's quite a nice UX imo.

Hi, where does you put this function? is it passed on message props? and what does this card mean? Is it refer to your component? Thanks!

**Comment by scriptndestroy** [comment]
This worked for me:
https://github.com/ReactTraining/history/issues/367#issuecomment-308860383
`const historyObject = createHistory({
    getUserConfirmation: (message, callback) => {
        callback(window.confirm(message));
    }
});

export default class AppRouter extends React.Component {
    render() {
        return (
            <Router history={historyObject}>
                <Routes {...appRoute} />
            </Router>
        );
    }
}`

**Comment by Rehre** [comment]
> I've been wrestling with the same issue, and finally found a solution that works for my use case. As others have pointed out, when a `POP` action triggers the `getUserConfirmation` method and the navigation is canceled, React Router correctly prevents the route transition within the React app, but the browser has already popped that item from `window.history`, so the URL in the browser will have changed.
> 
> My solution was to update my `getUserConfirmation` handler to detect a `POP` action and call `window.history.forward()` to return to the original URL within the browser history. But in order for `getUserConfirmation` to know what action was performed, I needed to pass along that information in the message from the `<Prompt />` component. The message is just a string, so I'm using a serialized JSON string to pass along the action along with my message and some other structured content that I need for my customized confirmation prompt.
> 
> As a bit of context, I'm using the `HashRouter` because my use case is within a browser extension. I can't verify whether the same issue happens or whether my solution is adequate for the `BrowserRouter`. I also don't know if this addresses all of the scenarios discussed in this thread, but I wanted to provide a reduced example of my solution in hopes that it can help someone out:
> 
> ```js-jsx
> // In the component that renders the <Prompt />:
> <Prompt
>   message={(_nextLocation, action) => {
>     return JSON.stringify({
>       action,
>       message: 'Some message'
>     })
>   }}
> />
> 
> // In the component that renders the router:
> <HashRouter
>   getUserConfirmation={(payload, callback) => {
>     const { action, message } = JSON.parse(payload);
> 
>     const confirmed = window.confirm(message)
>     
>     callback(confirmed)
> 
>     if (!confirmed && action === 'POP') window.history.forward()
>   }}
> >
>   ...
> </HashRouter />
> ```

your solution works for me too in BrowserRouter!, although i need to tweak a bit because i am not sure what is the reason but the prompt showed up again after forwarding the history, so what i did is to just check whether the next location has the same path or not (this might be too hacky so i am not sure if anyone should follow this) 


```js-jsx
  <BrowserRouter
    getUserConfirmation={(payload, callback) => {
      const { nextLocation, currentPath, action, message } = JSON.parse(
        payload
      );

      if (currentPath !== nextLocation.pathname) {
        const confirmed = window.confirm(message);
        callback(confirmed);

        if (!confirmed && action === 'POP') {
          window.history.forward();
        }
      }
    }}
  >
```

**Comment by pavelspichonak** [comment]
@wosephjeber @Andrew1431 your solutions work only when you click browser back button. But when you click browser forward button it works incorrectly because we call `window.history.forward();` but instead we need to call `window.history.back();`. Maybe you know how can we understand if back or forward browser button is clicked? Looks like for both we have `POP` action.

**Comment by mkrause** [comment]
@wosephjeber's solution worked for me, except I was getting double confirmation messages for some reason. Added a check to see if the location is actually different to resolve it:

```javascript
        <Prompt
          when={when}
          message={(locationNext, historyAction) => {
            return JSON.stringify({ historyAction, message, locationNext });
          }}
        />

      <BrowserRouter history={history}
        getUserConfirmation={(payload, callback) => {
          // https://github.com/ReactTraining/react-router/issues/5405#issuecomment-682730905
          const { historyAction, message, locationNext } = JSON.parse(payload);
          
          // Check if the current location and the next are equal. If so, ignore. This prevents some double
          // confirmation alerts.
          if (history.location.key === locationNext.key) {
            return;
          }
          
          const confirmed = window.confirm(message);
          
          callback(confirmed);
          
          if (!confirmed && historyAction === 'POP') {
            window.history.forward();
          }
        }}
      >
        <App/>
      </BrowserRouter>
```

**Comment by BrinsterAman** [comment]
I am also facing similar kind of issue, for me I am using `Prompt` with `history.push` and `browser back button`. When I click on either of them, and then click on `Cancel`, the url first gets changed. Then it gets back to previous url, but the `state` gets reset. Due to this, I lose my progress. Please help me to get through this.

**Comment by BrinsterAman** [comment]
I am also facing similar issue for both `history.push` and `browser back button`. Please help me to get through it.

**Comment by marcesdan** [comment]
5 years have passed. Don't use prompt

**Comment by radufilipescu** [comment]
@mkrause how did you fix the issue with "history" variable being undefined (as per [this](https://flaviocopes.com/react-router-uselocation-usehistory-undefined/) article) ?

**Comment by denchp** [comment]
From what I can tell this isn't a 'bug' per se.

There are two (or more) events that Prompt will capture - one of which is the `hashchange` event.  From what I've been able to reproduce the `hashchange` event is triggered _after_ navigation.  `Prompt` is then catching the hashchange event, using `window.confirm` to show the custom message.  If the user clicks cancel, `window.confirm` returns false and Prompt then returns us to our previous URL (hash).

From what I've been able to tell _there is no way around this_; `beforeunload` is not valid for hash changes because the page isn't actually unloaded (and it doesn't support custom messages), and `hashchange` is triggered _after_ the hash change event occurs - which results in the 'bug-like' behavior of the URL changing to the 'destination' url, then back to the 'original'.

This also triggers the components on the original view/page to re-render.

The only robust solution would be to store form/user/state data in a context (or redux store) and then restore it when the user navigates back, but that's a fairly heavy solution depending on the use case.

Anyone have any other ideas?

**Comment by cmonteiro128** [comment]
I am experience this issue as well with history "pop" actions (does not happen with push actions)

**Comment by tomvrampio** [comment]
Adding my experience with MUI and react-router-dom, just in case it helps anyone.

I was using MUI Buttons with hrefs. Prompt nearly worked, but was adding the new url before the user made their choice in Prompt, which is what led me here. I then used SufiyaanRajput's tweaked workaround until a colleague found a bug and forced me to reconsider. What fixed it for me was using MUI's component to use `react-router-dom` Link component i.e. `<Button component={Link} to="/whatever" ... />` - this meant I could use vanilla Prompt.

I think the difference is that the navigation event needs to go straight to `react-router-dom` rather than the browser.

**Comment by sriram15690** [comment]
> window.history.replaceState(null, null, "/your/current/path")

This reloads the component. If its a page with form, the user inputs vanish
